compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/TailRecursionCallsCollector.kt

```diff
@@ -73,18 +73,33 @@ fun collectTailRecursionCalls(irFunction: IrFunction): Set<IrCall> {
             expression.value.accept(this, valueKind)
         }
 
-        override fun visitContainerExpression(expression: IrContainerExpression, data: ElementKind) {
+        override fun visitExpressionBody(body: IrExpressionBody, data: ElementKind) =
+            body.acceptChildren(this, data)
+
+        override fun visitBlockBody(body: IrBlockBody, data: ElementKind) =
+            visitStatementContainer(body, data)
+
+        override fun visitContainerExpression(expression: IrContainerExpression, data: ElementKind) =
+            visitStatementContainer(expression, data)
+
+        private fun visitStatementContainer(expression: IrStatementContainer, data: ElementKind) {
             expression.statements.forEachIndexed { index, irStatement ->
-                val statementKind = if (index == expression.statements.lastIndex) {
+                val statementKind = when {
                     // The last statement defines the result of the container expression, so it has the same kind.
-                    data
-                } else {
-                    ElementKind.NOT_SURE
+                    index == expression.statements.lastIndex -> data
+                    // In a Unit-returning function, any statement directly followed by a `return` is a tail statement.
+                    isUnitReturn && expression.statements[index + 1].let {
```

Comment:
Could it be IndexOutOfBoundException?

Reply:
No, it's known that 0 <= index <= lastIndex due to `forEachIndexed`, and the lastIndex case is checked above.

Reply:
> the lastIndex case is checked above
oh..yes, indeed! Thank you!